File: android/app/src/main/java/com/david/tehilim/core/service/ChainShareLink.kt
package com.david.tehilim.core.service

import android.content.Context
import android.net.Uri
import com.david.tehilim.R
import com.david.tehilim.core.model.TehilimChain

/**
 * Encodage / décodage du lien de partage d'une **chaîne de Tehilim**.
 * Mirror de PrayerShareLink, mais le payload est minimal : seul l'**id de
 * chaîne** circule (le reste est lu depuis Supabase à l'ouverture).
 *
 * Lien = page de redirection `https://…/c/?id=…` (cliquable WhatsApp) qui
 * rouvre `tehilim://chain?id=…`. App Link sur le host + pathPrefix `/c/`.
 */
object ChainShareLink {

    const val SCHEME = "tehilim"
    const val HOST = "chain"
    const val WEB_BASE_URL = "https://tehilimapp.com/c/"

    /** Lien `https://…/c/?id=<chainId>`. */
    fun uri(chainId: String): Uri =
        Uri.parse(WEB_BASE_URL).buildUpon()
            .appendQueryParameter("id", chainId)
            .build()

    /** Message texte prêt à partager (WhatsApp) : description + lien. */
    fun shareMessage(context: Context, chain: TehilimChain): String {
        val link = uri(chain.id).toString()
        val prefix = context.getString(R.string.chain_share_prefix)
        val cta = context.getString(R.string.chain_share_cta)
        return "$prefix${chain.subjectLine}\n\n$cta\n$link"
    }

    /** True si l'URI ouvre une chaîne : `tehilim://chain` OU App Link `…/c/…`. */
    fun isChainLink(uri: Uri): Boolean {
        if (uri.scheme == SCHEME && uri.host == HOST) return true
        val p = uri.path
        return uri.scheme == "https" && (p == "/c" || p?.startsWith("/c/") == true)
    }

    /** Extrait l'id de chaîne (`?id=…`). Null si absent/invalide. */
    fun chainId(uri: Uri): String? {
        if (!isChainLink(uri)) return null
        return uri.getQueryParameter("id")?.trim()?.takeIf { it.isNotEmpty() }
    }
}

File: ios/Tehilim/Core/Services/ChainShareLink.swift
import Foundation

/// Encodage / décodage du lien de partage d'une **chaîne de Tehilim**.
///
/// Mirror de `PrayerShareLink`, mais le payload est minimal : seul l'**id de
/// chaîne** circule (le reste est lu depuis Supabase à l'ouverture). Le lien
/// est une page de redirection `https://…/c/?id=…` (cliquable dans WhatsApp),
/// qui rouvre `tehilim://chain?id=…`. Universal Link sur `/c/` (cf. AASA).
enum ChainShareLink {

    static let scheme = "tehilim"
    static let host = "chain"

    /// Page de redirection https (repo Pages séparé) — cliquable partout.
    static let webBaseURL = "https://tehilimapp.com/c/"

    // MARK: - Encodage

    /// Lien de partage `https://…/c/?id=<chainId>` représentant la chaîne.
    static func url(forChainId chainId: String) -> URL? {
        guard var comps = URLComponents(string: webBaseURL) else { return nil }
        comps.queryItems = [URLQueryItem(name: "id", value: chainId)]
        return comps.url
    }

    /// Message texte prêt à partager (WhatsApp) : description de la chaîne +
    /// le lien d'ouverture. Le destinataire tape le lien → ouvre l'app sur la
    /// chaîne et peut rejoindre / sélectionner ses Tehilim.
    static func shareMessage(for chain: TehilimChain) -> String {
        let fr = AppLocale.code == "fr"   // hébreu → message EN
        let link = url(forChainId: chain.id)?.absoluteString ?? ""
        var lines: [String] = []
        lines.append((fr ? "Chaîne de Tehilim — " : "Tehilim chain — ") + chain.subjectLine)
        lines.append("")
        lines.append(fr
            ? "Rejoins la chaîne et choisis les Tehilim que tu liras :"
            : "Join the chain and pick the Tehilim you'll read:")
        lines.append(link)
        return lines.joined(separator: "\n")
    }

    // MARK: - Décodage

    /// True si l'URL ouvre une chaîne : `tehilim://chain` OU Universal Link
    /// `https://…/c/…`. `URL.path` retire le « / » final → on accepte « /c »
    /// exact ET « /c/… » (sans matcher « /privacy » ni « /p »).
    static func isChainLink(_ url: URL) -> Bool {
        if url.scheme == scheme && url.host == host { return true }
        if url.scheme == "https" && (url.path == "/c" || url.path.hasPrefix("/c/")) { return true }
        return false
    }

    /// Extrait l'id de chaîne d'un lien (`?id=…`). Nil si absent/invalide.
    static func chainId(from url: URL) -> String? {
        guard isChainLink(url),
              let comps = URLComponents(url: url, resolvingAgainstBaseURL: false)
        else { return nil }
        let id = comps.queryItems?.first(where: { $0.name == "id" })?.value?
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard let id, !id.isEmpty else { return nil }
        return id
    }
}
